src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -4151,6 +4151,10 @@ bool StorageReplicatedMergeTree::optimize(
     const Names & deduplicate_by_columns,
     const Context & query_context)
 {
+    /// NOTE: exclusive lock cannot be used here, since this may lead to deadlock (see comments below),
+    /// but it should be safe to use non-exclusive to avoid dropping parts that may be required for processing queue.
+    auto table_lock = lockForShare(query_context.getCurrentQueryId(), query_context.getSettingsRef().lock_acquire_timeout);
```

Comment:
Oh, how we have lost it :(

Reply:
@alesapin OPTIMIZE should not normally require table-level lock.
The only issue is in waiting for entry to process.